Tencent Hunyuan Team: add HunyuanDiT related updates

gnobitab opened the pull request:
**This PR did the following things:**

1. Created `HunyuanDiTPipeline` in `src/diffusers/pipelines/hunyuandit/`  and `HunyuanDiT2DModel` in `./src/diffusers/models/transformers/`.
2. To support `HunyuanDiT2DModel`, added `HunyuanDiTBlock` and helper functions in `src/diffusers/models/attention.py` .
3. Uploaded the safetensors model to my huggingface: `XCLiu/HunyuanDiT-0523`
4. Tested the output of the migrated model+code is the same as our repo (https://github.com/Tencent/HunyuanDiT). Have tested different resolutions and batch sizes > 1 and made sure they work correctly.

In this branch, you can run HunyuanDiT in FP32 with:
```
python3 test_hunyuan_dit.py
``` 

which includes the following codes:
```py
import torch
from diffusers import HunyuanDiTPipeline

pipe = HunyuanDiTPipeline.from_pretrained("XCLiu/HunyuanDiT-0523", torch_dtype=torch.float32)
pipe.to('cuda')

### NOTE: HunyuanDiT supports both Chinese and English inputs
prompt = "一个宇航员在骑马"
#prompt = "An astronaut riding a horse"
image = pipe(height=1024, width=1024, prompt=prompt).images[0]

image.save("./img.png")
```

**Dependency:** 
maybe the `timm` package

TODO lists:
1. FP16 support: I didn't change the parameter `use_fp16` in `HunyuanDiTPipeline.__call__()`. The reason is `BertModel` does not support FP16 quantization. In our repo we only quantize the diffusion transformer to FP16. I guess there must be some smart way to support FP16.
2. Simplify and refactor the `HunyuanDiTBlock` related codes in `src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py`.
3. Refactor the pipeline and HunyuanDiT2DModel to diffusers style.
4. doc

Thank you so much! I'll be there and help with everything.

cc: @sayakpaul @yiyixuxu

sayakpaul commented on src/diffusers/models/attention.py:
```diff
@@ -84,6 +84,420 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
+    x_real, x_imag = x.float().reshape(*x.shape[:-1], -1, 2).unbind(-1)  # [B, S, H, D//2]
+    return torch.stack([-x_imag, x_real], dim=-1).flatten(3)
+
+def apply_rotary_emb(
```
I think it's better to have a `RotaryPositionEmbedding` class [here](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/embeddings.py) and use it across the board.

sayakpaul commented on src/diffusers/models/attention.py:
```diff
@@ -84,6 +84,420 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
+        x = x.permute(1, 0, 2)  # NLC -> LNC
+        x = torch.cat([x.mean(dim=0, keepdim=True), x], dim=0)  # (L+1)NC
+        x = x + self.positional_embedding[:, None, :].to(x.dtype)  # (L+1)NC
+        x, _ = F.multi_head_attention_forward(
```
We don't implementation attention classes like this. Instead. we leverage an "attention processor" mechanism. 

See here as an example how such an attention processor can be implemented:
https://github.com/huggingface/diffusers/blob/67b3fe0aaeddda19e8f61316a9b2fcceba4a4451/src/diffusers/models/attention_processor.py#L1340

And how it is leveraged from an attention implementation:
https://github.com/huggingface/diffusers/blob/67b3fe0aaeddda19e8f61316a9b2fcceba4a4451/src/diffusers/models/attention_processor.py#L539

sayakpaul commented on src/diffusers/models/attention.py:
```diff
@@ -84,6 +84,420 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
+        self.qdim = qdim
+        self.kdim = kdim
+        self.num_heads = num_heads
+        assert self.qdim % num_heads == 0, "self.qdim must be divisible by num_heads"
+        self.head_dim = self.qdim // num_heads
+        assert self.head_dim % 8 == 0 and self.head_dim <= 128, "Only support head_dim <= 128 and divisible by 8"
```
(nit): `raise ValueError`s in place of `assert`s.

sayakpaul commented on src/diffusers/models/attention.py:
```diff
@@ -84,6 +84,420 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
+        )
+        return x.squeeze(0)
+
+class HunyuanDiTCrossAttention(nn.Module):
```
Same comment as above regarding the implementation. It should be done with an attention processor class.

sayakpaul commented on src/diffusers/models/attention.py:
```diff
@@ -84,6 +84,420 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
+        assert self.head_dim % 8 == 0 and self.head_dim <= 128, "Only support head_dim <= 128 and divisible by 8"
+        self.scale = self.head_dim ** -0.5
+
+        self.q_proj = nn.Linear(qdim, qdim, bias=qkv_bias, **factory_kwargs)
+        self.kv_proj = nn.Linear(kdim, 2 * qdim, bias=qkv_bias, **factory_kwargs)
```
Also, we don't horizontally stack the attention projection matrices to promote readability in the library. So, this needs to be:

```python
self.q_proj = ...
self.k_proj = ...
self.v_proj = ...
```

Additionally, did you try repurposing the `Attention` class?
https://github.com/huggingface/diffusers/blob/67b3fe0aaeddda19e8f61316a9b2fcceba4a4451/src/diffusers/models/attention_processor.py#L43

sayakpaul commented on src/diffusers/models/attention.py:
```diff
@@ -84,6 +84,420 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
+        return out_tuple
+
+
+class HunyuanDiTAttention(nn.Module):
```
Same comment as above.

sayakpaul commented on src/diffusers/models/attention.py:
```diff
@@ -84,6 +84,420 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
+        #print('x:', hidden_states[0])
+
+        # 2. Cross-Attention
+        cross_inputs = (
+            self.norm3(hidden_states), encoder_hidden_states, freq_cis_img
+        )
+        hidden_states = hidden_states + self.attn2(*cross_inputs)[0]
+        #print('x:', hidden_states[0])
+
+        # FFN Layer ### NOTE: do not switch norm2 and norm3, otherwise will load wrong key when using pretrained model!
+        mlp_inputs = self.norm2(hidden_states)
```
This reads a little haphazard to me. Could you modify the state dictionary so that it's not the case?

gnobitab replied:
ok

sayakpaul commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,412 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2024 HunyuanDiT Authors and The HuggingFace Team. All rights reserved.
```

sayakpaul commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,412 @@
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+@dataclass
+class Transformer2DModelOutput(BaseOutput):
+    """
+    The output of [`Transformer2DModel`].
+
+    Args:
+        sample (`torch.Tensor` of shape `(batch_size, num_channels, height, width)` or `(batch size, num_vector_embeds - 1, num_latent_pixels)` if [`Transformer2DModel`] is discrete):
+            The hidden states output conditioned on the `encoder_hidden_states` input. If discrete, returns probability
+            distributions for the unnoised latent pixels.
+    """
+
+    sample: torch.Tensor
```
I don't think we need this. We can import it from here:
https://github.com/huggingface/diffusers/blob/67b3fe0aaeddda19e8f61316a9b2fcceba4a4451/src/diffusers/models/transformers/transformer_2d.py#L33

sayakpaul commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,412 @@
+import math
+from einops import repeat
+from timm.models.layers import to_2tuple
+class HunyuanDiTPatchEmbed(nn.Module):
```
This should go to https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/embeddings.py.
 
Also, can we leverage `PatchEmbed` here? 
https://github.com/huggingface/diffusers/blob/67b3fe0aaeddda19e8f61316a9b2fcceba4a4451/src/diffusers/models/embeddings.py#L124

If the modifications to support that are small, we'd prefer that I think.

sayakpaul commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,412 @@
+        x = self.norm(x)
+        return x
+
+def timestep_embedding(t, dim, max_period=10000, repeat_only=False):
```
Embeddings-related stuff should be in https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/embeddings.py. We should also try to leverage the existing utilities in the library as much as we can (if small modifications are needed, that is okay). For example this class:

https://github.com/huggingface/diffusers/blob/67b3fe0aaeddda19e8f61316a9b2fcceba4a4451/src/diffusers/models/embeddings.py#L190

sayakpaul commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,412 @@
+
+    Inherit ModelMixin and ConfigMixin to be compatible with the sampler StableDiffusionPipeline of diffusers.
+
+    Parameters
+    ----------
+    args: argparse.Namespace
+        The arguments parsed by argparse.
+    input_size: tuple
+        The size of the input image.
+    patch_size: int
+        The size of the patch.
+    in_channels: int
+        The number of input channels.
+    hidden_size: int
+        The hidden size of the transformer backbone.
+    depth: int
+        The number of transformer blocks.
+    num_heads: int
+        The number of attention heads.
+    mlp_ratio: float
+        The ratio of the hidden size of the MLP in the transformer block.
+    log_fn: callable
+        The logging function.
```
(nit): needs to be updated.

sayakpaul commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,412 @@
+        self.num_heads = num_attention_heads
+        self.hidden_size = hidden_size
+        self.text_states_dim = text_dim
+        self.text_states_dim_t5 = 2048
+        self.text_len = 77
+        self.text_len_t5 = 256 ### NOTE: These numbers are hardcoded for now, seems will not change in near future
```
We should still pass it from `__init__` and then assign them from the constructor here.

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1120 @@
+            A scheduler to be used in combination with HunyuanDiT to denoise the encoded image latents.
+    """
+
+    model_cpu_offload_seq = "text_encoder->embedder_t5->tokenizer->tokenizer_t5->transformer->vae"
```
Tokenizer can be excluded from here.

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1120 @@
+            progress_bar_config: Dict[str, Any] = None,
+            embedder_t5=T5EncoderModel,
+            tokenizer_t5=MT5Tokenizer,
+            infer_mode='torch',
```
What is `infer_mode` used for? We usually don't allow such variables from the pipeline inits.

gnobitab replied:
it is used for swtiching between TensorRT and PyTorch. I'll remove it for now and we can support TensorRT later.

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1120 @@
+        super().__init__()
+
+        # ========================================================
+        self.infer_mode = infer_mode
+
+        # ========================================================
+        if progress_bar_config is None:
+            progress_bar_config = {}
+        if not hasattr(self, '_progress_bar_config'):
+            self._progress_bar_config = {}
+        self._progress_bar_config.update(progress_bar_config)
+        # ========================================================
```
We don't configure progressbar from pipeline inits.

yiyixuxu replied:
+1 
maybe just add a `_progress_bar_config` attribute to the pipeline if you need special config

gnobitab replied:
I'll keep the progress bar simple as StableDiffusionPipeline

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1120 @@
+            embedder_t5=embedder_t5,
+        )
+
+        self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
```
Can you elaborate what issue you're facing here? This is anti-pattern in the library.

gnobitab replied:
There is an extra 'pooler' module initialized in BertModel. Don't know how to configure correctly to remove it and fit in `from_pretrained`. In our pretrained Bert, we turned this layer off at intialization.

yiyixuxu replied:
ohh I see https://github.com/huggingface/transformers/blob/965e98dc54c4853b34b3a49810f24856a9fcdba8/src/transformers/models/bert/modeling_bert.py#L976

@sayakpaul can you look into how we can make it work with `from_pretrained()`?
do we need to make a wrapper of bert model inside diffusers?

sayakpaul replied:
Could we do something like this when passing the `BertModel`?

```python
from transformers import BertModel

bert_ckpt = "..."
bert_model = BertModel.from_pretrained(bert_ckpt, add_pooling_layer=False)
```

This way, we should be able to obtain the right variant of the `BertModel` that we want and serialize it along with the other components of the pipeline.

sayakpaul replied:
Doesn't this work?

```python
from transformers import BertModel
from diffusers import HunyuanDiTPipeline

ckpt_id = "XCLiu/HunyuanDiT-0523"
model = BertModel.from_pretrained(ckpt_id, add_pooling_layer=False, subfolder="text_encoder")

pipeline = HunyuanDiTPipeline.from_pretrained(
    ckpt_id, text_encoder=model, ...
)
```

@gnobitab could you give this a try without the manual `to_empty()` operation?

sayakpaul replied:
Just a disclaiming that when you load `BertModel` like the above, you would notice:

```bash
Some weights of the model checkpoint at XCLiu/HunyuanDiT-0523 were not used when initializing BertModel: ['pooler.dense.bias', 'pooler.dense.weight']
```

You mentioned the `pooler` layers shouldn't matter. So, LMK.

yiyixuxu replied:
we should have this work 
```python
pipeline = HunyuanDiTPipeline.from_pretrained(ckpt_id)
```

sayakpaul replied:
Yes, this would work after the pipeline is pushed with the modifications I suggested.

```python
from transformers import BertModel
from diffusers import HunyuanDiTPipeline

ckpt_id = "XCLiu/HunyuanDiT-0523"
model = BertModel.from_pretrained(ckpt_id, add_pooling_layer=False, subfolder="text_encoder")

pipeline = HunyuanDiTPipeline.from_pretrained(
    ckpt_id, text_encoder=model, ...
)
pipeline.push_to_hub(...)
```

And then  

```python
pipeline = HunyuanDiTPipeline.from_pretrained(ckpt_id)
```

should work without `to_empty()`.

yiyixuxu replied:
ohhh ok!

gnobitab replied:
Thanks for the suggestion! This code snippet works without `to_empty`. 
But after I saved the pipeline to with 
```py
from transformers import BertModel
bert_model = BertModel.from_pretrained("XCLiu/HunyuanDiT-0523", add_pooling_layer=False, subfolder="text_encoder")

pipe = HunyuanDiTPipeline.from_pretrained("XCLiu/HunyuanDiT-0523", text_encoder=bert_model, transformer=model, torch_dtype=torch.float32)
pipe.to('cuda')
pipe.save_pretrained("../HunyuanDiT-ckpt")
``` 
and reload it with
```py
pipeline = HunyuanDiTPipeline.from_pretrained("../HunyuanDiT-ckpt")
pipe.to('cuda')
```
The same error appears again...
<img width="1347" alt="image" src="https://github.com/huggingface/diffusers/assets/1157982/f69c2041-a25f-4dd6-b0df-a63681cbefcd">

sayakpaul replied:
I am able to successfully do:

```python
from transformers import BertModel
from diffusers import HunyuanDiTPipeline, HunyuanDiT2DModel
from huggingface_hub import hf_hub_download
import safetensors
import torch

def load_transformer():
    model_config = HunyuanDiT2DModel.load_config("XCLiu/HunyuanDiT-0523", subfolder="transformer")
    model = HunyuanDiT2DModel.from_config(model_config)

    ckpt_path = hf_hub_download(
        "XCLiu/HunyuanDiT-0523",
        filename ="diffusion_pytorch_model.safetensors",
        subfolder="transformer",
    )
    state_dict = safetensors.torch.load_file(ckpt_path)

    num_layers = 40
    for i in range(num_layers):
        
        # attn1
        # Wkqv -> to_q, to_k, to_v
        q, k, v = torch.chunk(state_dict[f"blocks.{i}.attn1.Wqkv.weight"], 3, dim=0)
        q_bias, k_bias, v_bias = torch.chunk(state_dict[f"blocks.{i}.attn1.Wqkv.bias"], 3, dim=0)
        state_dict[f"blocks.{i}.attn1.to_q.weight"] = q
        state_dict[f"blocks.{i}.attn1.to_q.bias"] = q_bias
        state_dict[f"blocks.{i}.attn1.to_k.weight"] = k
        state_dict[f"blocks.{i}.attn1.to_k.bias"] = k_bias
        state_dict[f"blocks.{i}.attn1.to_v.weight"] = v
        state_dict[f"blocks.{i}.attn1.to_v.bias"] = v_bias
        state_dict.pop(f"blocks.{i}.attn1.Wqkv.weight")
        state_dict.pop(f"blocks.{i}.attn1.Wqkv.bias")
        
        # q_norm, k_norm -> norm_q, norm_k
        state_dict[f"blocks.{i}.attn1.norm_q.weight"] = state_dict[f"blocks.{i}.attn1.q_norm.weight"]
        state_dict[f"blocks.{i}.attn1.norm_q.bias"] = state_dict[f"blocks.{i}.attn1.q_norm.bias"]
        state_dict[f"blocks.{i}.attn1.norm_k.weight"] = state_dict[f"blocks.{i}.attn1.k_norm.weight"]
        state_dict[f"blocks.{i}.attn1.norm_k.bias"] = state_dict[f"blocks.{i}.attn1.k_norm.bias"]

        state_dict.pop(f"blocks.{i}.attn1.q_norm.weight")
        state_dict.pop(f"blocks.{i}.attn1.q_norm.bias")
        state_dict.pop(f"blocks.{i}.attn1.k_norm.weight")
        state_dict.pop(f"blocks.{i}.attn1.k_norm.bias")

        # out_proj -> to_out
        state_dict[f"blocks.{i}.attn1.to_out.0.weight"] = state_dict[f"blocks.{i}.attn1.out_proj.weight"]
        state_dict[f"blocks.{i}.attn1.to_out.0.bias"] = state_dict[f"blocks.{i}.attn1.out_proj.bias"]
        state_dict.pop(f"blocks.{i}.attn1.out_proj.weight")
        state_dict.pop(f"blocks.{i}.attn1.out_proj.bias")

        # attn2
        # kq_proj -> to_k, to_v
        k, v = torch.chunk(state_dict[f"blocks.{i}.attn2.kv_proj.weight"], 2, dim=0)
        k_bias, v_bias = torch.chunk(state_dict[f"blocks.{i}.attn2.kv_proj.bias"], 2, dim=0)
        state_dict[f"blocks.{i}.attn2.to_k.weight"] = k
        state_dict[f"blocks.{i}.attn2.to_k.bias"] = k_bias
        state_dict[f"blocks.{i}.attn2.to_v.weight"] = v
        state_dict[f"blocks.{i}.attn2.to_v.bias"] = v_bias
        state_dict.pop(f"blocks.{i}.attn2.kv_proj.weight")
        state_dict.pop(f"blocks.{i}.attn2.kv_proj.bias")
        
        # q_proj -> to_q
        state_dict[f"blocks.{i}.attn2.to_q.weight"] = state_dict[f"blocks.{i}.attn2.q_proj.weight"]
        state_dict[f"blocks.{i}.attn2.to_q.bias"] = state_dict[f"blocks.{i}.attn2.q_proj.bias"]
        state_dict.pop(f"blocks.{i}.attn2.q_proj.weight")
        state_dict.pop(f"blocks.{i}.attn2.q_proj.bias")
        
        # q_norm, k_norm -> norm_q, norm_k
        state_dict[f"blocks.{i}.attn2.norm_q.weight"] = state_dict[f"blocks.{i}.attn2.q_norm.weight"]
        state_dict[f"blocks.{i}.attn2.norm_q.bias"] = state_dict[f"blocks.{i}.attn2.q_norm.bias"]
        state_dict[f"blocks.{i}.attn2.norm_k.weight"] = state_dict[f"blocks.{i}.attn2.k_norm.weight"]
        state_dict[f"blocks.{i}.attn2.norm_k.bias"] = state_dict[f"blocks.{i}.attn2.k_norm.bias"]

        state_dict.pop(f"blocks.{i}.attn2.q_norm.weight")
        state_dict.pop(f"blocks.{i}.attn2.q_norm.bias")
        state_dict.pop(f"blocks.{i}.attn2.k_norm.weight")
        state_dict.pop(f"blocks.{i}.attn2.k_norm.bias")

        # out_proj -> to_out
        state_dict[f"blocks.{i}.attn2.to_out.0.weight"] = state_dict[f"blocks.{i}.attn2.out_proj.weight"]
        state_dict[f"blocks.{i}.attn2.to_out.0.bias"] = state_dict[f"blocks.{i}.attn2.out_proj.bias"]
        state_dict.pop(f"blocks.{i}.attn2.out_proj.weight")
        state_dict.pop(f"blocks.{i}.attn2.out_proj.bias")

        # switch norm 2 and norm 3
        norm2_weight = state_dict[f"blocks.{i}.norm2.weight"]
        norm2_bias = state_dict[f"blocks.{i}.norm2.bias"]
        state_dict[f"blocks.{i}.norm2.weight"] = state_dict[f"blocks.{i}.norm3.weight"]
        state_dict[f"blocks.{i}.norm2.bias"] = state_dict[f"blocks.{i}.norm3.bias"]
        state_dict[f"blocks.{i}.norm3.weight"] = norm2_weight
        state_dict[f"blocks.{i}.norm3.bias"] = norm2_bias

    model.load_state_dict(state_dict)
    return model


bert_model = BertModel.from_pretrained(
    "XCLiu/HunyuanDiT-0523", add_pooling_layer=False, subfolder="text_encoder"
)
model = load_transformer()
pipe = HunyuanDiTPipeline.from_pretrained("XCLiu/HunyuanDiT-0523", text_encoder=bert_model, transformer=model)
pipe.save_pretrained("HunyuanDiT-ckpt")

del pipe 

pipe = HunyuanDiTPipeline.from_pretrained("HunyuanDiT-ckpt")
```

With the following change:

```diff
diff --git a/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py b/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
index 381ded58..f373826e 100644
--- a/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
+++ b/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
@@ -408,7 +408,7 @@ class HunyuanDiTPipeline(DiffusionPipeline, TextualInversionLoaderMixin, LoraLoa
             embedder_t5=embedder_t5,
         )
 
-        self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
+        # self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
 
         if safety_checker is None and requires_safety_checker:
             logger.warning(

```

LMK if that works?

gnobitab replied:
It doesn't work for me ... I copy-paste your test code and it still shows this:

<img width="1363" alt="image" src="https://github.com/huggingface/diffusers/assets/1157982/59f9dfa3-cb09-4d0c-9eeb-708b9705d716">

So (1) Sayak could you share your package versions? (2) could you try `pipe.to('cuda')` after `pipe = HunyuanDiTPipeline.from_pretrained("HunyuanDiT-ckpt")`. It's `pipe.to('cuda')` that pops errors. Thanks.

sayakpaul replied:
I think you didn't apply the patch I mentioned in my comment. Here it is again for your reference: 

```diff
diff --git a/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py b/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
index 381ded58..f373826e 100644
--- a/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
+++ b/src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py
@@ -408,7 +408,7 @@ class HunyuanDiTPipeline(DiffusionPipeline, TextualInversionLoaderMixin, LoraLoa
             embedder_t5=embedder_t5,
         )
 
-        self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
+        # self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
 
         if safety_checker is None and requires_safety_checker:
             logger.warning(
```

gnobitab replied:
it looks like this patch only make `self.text_encoder.pooler.to_empty(device='cpu')` a comment? if so, I already did that.

sayakpaul replied:
That is quite weird then. 

Can you check if you are on the latest version of `transformers` and `accelerate`?

gnobitab replied:
I'm using `transformers==0.41.1` and `accelerate==0.30.1`. let me find out what's wrong. it is weird.

gnobitab replied:
I finally get it to work! With the following codes:

```py
from transformers import BertModel
bert_model = BertModel.from_pretrained("XCLiu/HunyuanDiT-0523", add_pooling_layer=True, subfolder="text_encoder")

pipe = HunyuanDiTPipeline.from_pretrained("XCLiu/HunyuanDiT-0523", text_encoder=bert_model, transformer=model, torch_dtype=torch.float32)
pipe.to('cuda')
pipe.save_pretrained("HunyuanDiT-ckpt")

del pipe

pipe = HunyuanDiTPipeline.from_pretrained("HunyuanDiT-ckpt")
pipe.cuda()
```

In this way, the random-initialized weights in the 'pooler' layer with be stored, but it will not affect the results and kept the code snippet simple as `pipe = HunyuanDiTPipeline.from_pretrained('xxx')`. Thank you Sayak for the help!

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1120 @@
+
+        self.text_encoder.pooler.to_empty(device='cpu') ### workaround for the meta device in pooler...
+
+        if hasattr(scheduler.config, "steps_offset") and scheduler.config.steps_offset != 1:
+            deprecation_message = (
+                f"The configuration file of this scheduler: {scheduler} is outdated. `steps_offset`"
+                f" should be set to 1 instead of {scheduler.config.steps_offset}. Please make sure "
+                "to update the config accordingly as leaving `steps_offset` might led to incorrect results"
+                " in future versions. If you have downloaded this checkpoint from the Hugging Face Hub,"
+                " it would be very nice if you could open a Pull request for the `scheduler/scheduler_config.json`"
+                " file"
+            )
+            deprecate("steps_offset!=1", "1.0.0", deprecation_message, standard_warn=False)
+            new_config = dict(scheduler.config)
+            new_config["steps_offset"] = 1
+            scheduler._internal_dict = FrozenDict(new_config)
```
This is a new pipeline so we don't need this deprecation :)

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1120 @@
+            new_config["steps_offset"] = 1
+            scheduler._internal_dict = FrozenDict(new_config)
+
+        if hasattr(scheduler.config, "clip_sample") and scheduler.config.clip_sample is True:
```
This is a new pipeline so we don't need this deprecation :)

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1120 @@
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+        self.register_to_config(requires_safety_checker=requires_safety_checker)
+
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline._encode_prompt
```
We should leverage this method instead:
https://github.com/huggingface/diffusers/blob/67b3fe0aaeddda19e8f61316a9b2fcceba4a4451/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L302

and completely get rid of `_encode_prompt()`.

gnobitab replied:
Deleted. `_encode_prompt` is not used in the whole pipeline.

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1120 @@
+
+        return prompt_embeds, negative_prompt_embeds, attention_mask, uncond_attention_mask
+
+    def _convert_to_rgb(self, image):
+        return image.convert('RGB')
```
This can be done using the `VaeImageProcessor`:
https://github.com/huggingface/diffusers/blob/67b3fe0aaeddda19e8f61316a9b2fcceba4a4451/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L267

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1120 @@
+        img_for_clip = image_preprocess(img)
+        
+        img_for_clip = img_for_clip.unsqueeze(0)
+        img_clip_embedding = self.img_encoder(img_for_clip.to(device)).to(dtype=torch.float16)
```
Where is this `img_encoder` coming from? Didn't see it in the pipeline init.

gnobitab replied:
I think this `encode_img` function is to support image-to-image generation later. Safe to delete for now.

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1120 @@
+            )
+        return image, has_nsfw_concept
+
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.decode_latents
+    def decode_latents(self, latents):
```
No need to leverage deprecated utilities in new pipelines.

gnobitab replied:
Deleted.

gnobitab commented on src/diffusers/models/attention.py:
```diff
@@ -84,6 +84,420 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
+def modulate(x, shift, scale):
+    return x * (1 + scale.unsqueeze(1)) + shift.unsqueeze(1)
+
+class FP32_Layernorm(nn.LayerNorm):
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        origin_dtype = inputs.dtype
+        return F.layer_norm(inputs.float(), self.normalized_shape, self.weight.float(), self.bias.float(),
+                            self.eps).to(origin_dtype)
+
+
+class FP32_SiLU(nn.SiLU):
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        return torch.nn.functional.silu(inputs.float(), inplace=False).to(inputs.dtype)
```
they cause numerical errors when quantizing to FP16, so we intentionally do this type conversion for FP16 inference

gnobitab commented on src/diffusers/models/attention.py:
```diff
@@ -84,6 +84,420 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
 
+### TODO: XCLiu: some ugly helper functions, please clean later
+### ==== begin ====
+def modulate(x, shift, scale):
+    return x * (1 + scale.unsqueeze(1)) + shift.unsqueeze(1)
```
thanks for pointing out! Let's fix it later

gnobitab commented on src/diffusers/models/attention.py:
```diff
@@ -84,6 +84,420 @@ def forward(self, x: torch.Tensor, objs: torch.Tensor) -> torch.Tensor:
+        AssertionError: If the target tensor 'x' doesn't have the expected number of dimensions.
+    """
+    ndim = x.ndim
+    assert 0 <= 1 < ndim
```
will fix.

yiyixuxu reviewed (commented):
Thanks for the PR!
I think the first a few main tasks are:
1. rewrite attentions using diffusers (I will help with this)
2. remove `timm` dependency (I think you only use it for a simple function, it should be easy to remove)
3. refactor the pipelines (I agree with @sayakpaul's comments there, we can start with these things he has pointed out)

yiyixuxu commented:
Hi:

I removed `HunyuanDiTAttention` and `HunyuanDiTCrossAttention` to use our `Attention` class with a `HunyuanAttnProcessor2_0` attention processor instead, 

feel free to test out the [PR branch](https://github.com/huggingface/diffusers/pull/8265) and cherry-pick this commit https://github.com/huggingface/diffusers/pull/8265/commits/3f85b1d257a9184b38772fa54997d334da1e1fae if results are ok to you

gnobitab commented:
Hi I did the following things:
1. cleaned the pipeline/transformer codes according to Sayak's suggestions. 
2. switched to yiyi's refactored attention. 
3. new test code in `test_hunyuan_dit.py`. based on yiyi's test code, switched norm2 and norm3, fixed generator to seed 0. the image should be:

![image](https://github.com/huggingface/diffusers/assets/1157982/2b14c4d0-7739-4304-a8a0-78ccdb760dab)

For now, I will not change the remote state_dict `XCLiu/HunyuanDiT-0523`. will update after we finished everything.

Please review and comment, thx!

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1007 @@
+    rope = get_2d_rotary_pos_embed(head_size, *sub_args)
+    return rope
+
+def rescale_noise_cfg(noise_cfg, noise_pred_text, guidance_rescale=0.0):
```
I think this is copied from here?
https://github.com/huggingface/diffusers/blob/cef4a512238c5f1cd88c4890e344d64459f92bac/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L59

So, let's add the "# Copied from ..." statement.

sayakpaul commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1007 @@
+        # 7. Prepare extra step kwargs. TODO: Logic should ideally just be moved out of the pipeline
+        extra_step_kwargs = self.prepare_extra_step_kwargs(generator, eta)
+        
+        # ========================================================================
+        # Arguments: style. (A fixed argument. Don't Change it.)
+        # ========================================================================
+        style = torch.as_tensor([0, 0] * batch_size, device=self._execution_device)
+
+        # ========================================================================
+        # Inner arguments: image_meta_size (Please refer to SDXL.)
+        # ========================================================================
+        src_size_cond = (1024, 1024)
+        if isinstance(src_size_cond, int):
+            src_size_cond = [src_size_cond, src_size_cond]
+        if not isinstance(src_size_cond, (list, tuple)):
+            raise TypeError(f"`src_size_cond` must be a list or tuple, but got {type(src_size_cond)}")
+        if len(src_size_cond) != 2:
+            raise ValueError(f"`src_size_cond` must be a tuple of 2 integers, but got {len(src_size_cond)}")
+        size_cond = list(src_size_cond) + [target_width, target_height, 0, 0]
+        image_meta_size = torch.as_tensor([size_cond] * 2 * batch_size, device=self._execution_device)
```
I see a number of magic values here such as 1024, [0, 0], etc. Could we please place them as the call arguments of the pipeline?

sayakpaul reviewed (commented):
Thanks for the changes and for incorporating our feedback so nicely so far! I have a left a couple of comments. 

Broadly, I think, for the next phase of the integration, we could focus on:

* Solving the `BertModel` pooler problem as mentioned in the comments. 
* Remove all the unnecessary blocks related to attention, etc. 
* Move the embedding related functionalities to `embeddings.py`. 
* Try to have Rotary Embeddings as a class. 

Would like to also cross-check with @yiyixuxu here at this point if these points make sense.

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1007 @@
+        # ========================================================================
+        # Inner arguments: image_meta_size (Please refer to SDXL.)
+        # ========================================================================
+        src_size_cond = (1024, 1024)
```
1. is this meant to be variables instead of constant here? 
2. is this same as `original_size` in SDXL? https://github.com/huggingface/diffusers/blob/0ab63ff6478b7cc6b5ae0d46c7c386d476cfa87f/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L858

gnobitab replied:
I guess they are the same. I make them constants now. Sayak wants them to be in the __call__ arguments because they look like magic numbers. @sayakpaul shall we just simply make it constant?

sayakpaul replied:
If the behaviour is same as SDXL, they should be pipeline call arguments as evident from here:

https://github.com/huggingface/diffusers/blob/0ab63ff6478b7cc6b5ae0d46c7c386d476cfa87f/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L858

Let me know if there's anything unclear.

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1007 @@
+        print(negative_prompt)
+
+        # 2. Calculate neccessary elements for HunyuanDiT
+        target_height = int((height // 16) * 16)
```
is this same as `target_size` argument in SDXL? https://github.com/huggingface/diffusers/blob/0ab63ff6478b7cc6b5ae0d46c7c386d476cfa87f/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L860

if so, can we handle it similarly as how it's done in SDXL? use the same name, accept the same values etc

gnobitab replied:
they are not the same. target_height and target_width are derived from height and width argument. SDXL has height and width argument too. I changed the height and width to the same as SDXL in __call__ arguments and renamed target_height to rounded_height to be more accurate.

gnobitab replied:
I found these target_height and target_width are the same as width and height...I directly replaced them with height and width

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1007 @@
+                # expand scalar t to 1-D tensor to match the 1st dim of latent_model_input
+                t_expand = torch.tensor([t] * latent_model_input.shape[0], device=latent_model_input.device)
+
+                if use_fp16:
+                    latent_model_input = latent_model_input.half()
+                    t_expand = t_expand.half()
+                    prompt_embeds = prompt_embeds.half()
+                    ims = image_meta_size.half() if image_meta_size is not None else None
+                else:
+                    ims = image_meta_size if image_meta_size is not None else None
```
don't think it is needed for the most part, e.g. if the pipeline is loaded as fp16, the `latent_model_input` should already be in fp16, we can also explicitly set the dtype inside like this whenever it is needed https://github.com/huggingface/diffusers/blob/0ab63ff6478b7cc6b5ae0d46c7c386d476cfa87f/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L470
```suggestion
                ims = image_meta_size if image_meta_size is not None else None
```

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1007 @@
+            style: Optional[torch.LongTensor] = None,
+            use_fp16: bool = False,
+            freqs_cis_img: Optional[tuple] = None,
+            learn_sigma: bool = True,
```
Do we need this config for the pipeline? It should depend on whether the model predicts variance too, no?
so if the pipeline needs to work with both type of models, we can make this a model config instead; if the pipeline only works with one type of model, we do not need this argument

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1007 @@
+                    return_dict=False,
+                )
+
+                if learn_sigma:
```
possible to remove this condition? or use model config instead

gnobitab replied:
removed

yiyixuxu reviewed (commented):
thanks!
looks much better now! I left a few more comments. 
remaining to-dos:
1. can you address the comments on the pipeline?
2. I will help refactor models
3. working on tests (cc @sayakpaul here, can you give more guidance)
4. make the `BertModel` work with `DiffusionPipeline.from_pretrained()` cc @sayakpaul and @DN6 here  see https://github.com/huggingface/diffusers/pull/8240/files#r1611508812

gnobitab commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1007 @@
+        return x
+
+
+def get_fill_resize_and_crop(src, tgt):   
```
it is different. `_resize_and_crop` takes PIL Image as input and outputs another PIL Image. Our method takes two tuples as input and outputs two tuples.

gnobitab replied:
So I don't think it's appropriate to put it in `image` processor either... I renamed this function instead, but please check

gnobitab replied:
I renamed to `get_fill_resize_and_crop_tuple` ... not sure how to make this better

sayakpaul commented:
> Our BertModel cannot work in FP16. So to make our pipeline to FP16, we have to make only the transformer to FP16 instead of just pipe = HunyuanDiTPipeline.from_pretrained("XCLiu/HunyuanDiT-0523", transformer=model, torch_dtype=torch.float16)

Why wouldn't it work, though? Could you provide more details here? In order for the pipeline to operate in `torch.float16` all the components need to be in `torch.float16`, or we have deal with it differently like so:

https://github.com/huggingface/diffusers/blob/b3d10d6d65a80593627c6738fbeded2f69b5129f/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L1264

Additionally, could you see if [this](https://github.com/huggingface/diffusers/pull/8240/files#r1614280384) worked so that we can avoid `self.text_encoder.pooler.to_empty(device='cpu')`?

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1010 @@
+#################################################################################
+# https://github.com/facebookresearch/mae/blob/main/util/pos_embed.py
+
+def get_2d_sincos_pos_embed(embed_dim, start, *args, cls_token=False, extra_tokens=0):
+    """
+    grid_size: int of the grid height and width
+    return:
+    pos_embed: [grid_size*grid_size, embed_dim] or [1+grid_size*grid_size, embed_dim] (w/ or w/o cls_token)
+    """
+    grid = get_meshgrid(start, *args)   # [2, H, w]
+    # grid_h = np.arange(grid_size, dtype=np.float32)
+    # grid_w = np.arange(grid_size, dtype=np.float32)
+    # grid = np.meshgrid(grid_w, grid_h)  # here w goes first
+    # grid = np.stack(grid, axis=0)   # [2, W, H]
+
+    grid = grid.reshape([2, 1, *grid.shape[1:]])
+    pos_embed = get_2d_sincos_pos_embed_from_grid(embed_dim, grid)
+    if cls_token and extra_tokens > 0:
+        pos_embed = np.concatenate([np.zeros([extra_tokens, embed_dim]), pos_embed], axis=0)
+    return pos_embed
```
this is not used, no?
```suggestion
```

gnobitab replied:
no, removed

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1010 @@
+    return pos_embed
+
+
+def get_2d_sincos_pos_embed_from_grid(embed_dim, grid):
+    assert embed_dim % 2 == 0
+
+    # use half of dimensions to encode grid_h
+    emb_h = get_1d_sincos_pos_embed_from_grid(embed_dim // 2, grid[0])  # (H*W, D/2)
+    emb_w = get_1d_sincos_pos_embed_from_grid(embed_dim // 2, grid[1])  # (H*W, D/2)
+
+    emb = np.concatenate([emb_h, emb_w], axis=1)    # (H*W, D)
+    return emb
```
```suggestion
```

gnobitab replied:
not used, removed

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1010 @@
+    return emb
+
+
+def get_1d_sincos_pos_embed_from_grid(embed_dim, pos):
+    """
+    embed_dim: output dimension for each position
+    pos: a list of positions to be encoded: size (W,H)
+    out: (M, D)
+    """
+    assert embed_dim % 2 == 0
+    omega = np.arange(embed_dim // 2, dtype=np.float64)
+    omega /= embed_dim / 2.
+    omega = 1. / 10000**omega  # (D/2,)
+
+    pos = pos.reshape(-1)  # (M,)
+    out = np.einsum('m,d->md', pos, omega)  # (M, D/2), outer product
+
+    emb_sin = np.sin(out)   # (M, D/2)
+    emb_cos = np.cos(out)   # (M, D/2)
+
+    emb = np.concatenate([emb_sin, emb_cos], axis=1)  # (M, D)
+    return emb
```
```suggestion
```

gnobitab replied:
no, removed

yiyixuxu commented on src/diffusers/models/attention_processor.py:
```diff
@@ -2697,3 +2808,84 @@ def __call__(
+        xq_out = (xq.float() * cos + rotate_half(xq.float()) * sin).type_as(xq)
+        if xk is not None:
+            xk_out = (xk.float() * cos + rotate_half(xk.float()) * sin).type_as(xk)
+    else:
+        xq_ = torch.view_as_complex(xq.float().reshape(*xq.shape[:-1], -1, 2))  # [B, S, H, D//2]
+        freqs_cis = reshape_for_broadcast(freqs_cis, xq_, head_first).to(xq.device)   # [S, D//2] --> [1, S, 1, D//2]
+        xq_out = torch.view_as_real(xq_ * freqs_cis).flatten(3).type_as(xq)
+        if xk is not None:
+            xk_ = torch.view_as_complex(xk.float().reshape(*xk.shape[:-1], -1, 2))  # [B, S, H, D//2]
+            xk_out = torch.view_as_real(xk_ * freqs_cis).flatten(3).type_as(xk)
```
I think we have control whether `freqs_cis` is a tuple or not, so let's simplify the code here 
```suggestion
```

gnobitab replied:
make sense, removed

yiyixuxu commented:
I refactored the model here https://github.com/huggingface/diffusers/pull/8310
you can see the changed I made in this commit https://github.com/huggingface/diffusers/pull/8310/commits/b0e0da28c4d0f057824faacb23da6b06dd43a786

I did below things:

1. refactor `HunyuanDiT2DModel`: removed `HunyuanDiTPatchEmbed`, `HunyuanDiTTimestepEmbedder`
2. refactor` apply_rotary_emb`: moved all the `get_2d_rotary_pos_embed_*` functions and `apply_rotary_emb` to `embeddings.py`
3 moved `HunyuanDiTBlock` and other Hunyuan-specific block to the same file as `HunyuanDiT2DModel`

I changed arg names to be more aligned with our transformer models and blocks also removed bunch of functionalities that are not used in this implementation - let me know if any of these does not make sense!

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1010 @@
+            cross_attention_kwargs: Optional[Dict[str, Any]] = None,
+            guidance_rescale: float = 0.0,
+            image_meta_size: Optional[torch.LongTensor] = None,
+            **kwargs
```
```suggestion
```

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1010 @@
+                "not-safe-for-work" (nsfw) content.
+        """
+        
+        callback = kwargs.pop("callback", None)
+        callback_steps = kwargs.pop("callback_steps", None)
+
+        if callback is not None:
+            deprecate(
+                "callback",
+                "1.0.0",
+                "Passing `callback` as an input argument to `__call__` is deprecated, consider use `callback_on_step_end`",
+            )
+        if callback_steps is not None:
+            deprecate(
+                "callback_steps",
+                "1.0.0",
+                "Passing `callback_steps` as an input argument to `__call__` is deprecated, consider use `callback_on_step_end`",
+            )
```
we don't need this for new pipeline!
```suggestion
```

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,1010 @@
+        # ========================================================================
+        # Inner arguments: image_meta_size (Please refer to SDXL.)
+        # ========================================================================
+        size_cond = [1024, 1024] + [width, height, 0, 0]
```
idealy we do the same thing as SDXL with this method 
https://github.com/huggingface/diffusers/blob/0ab63ff6478b7cc6b5ae0d46c7c386d476cfa87f/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L725

1. we have pipeline argument `original_size` and `target_size`
2. default them to (`height`, `width`) https://github.com/huggingface/diffusers/blob/0ab63ff6478b7cc6b5ae0d46c7c386d476cfa87f/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L1039
3.  we should probably rename `image_meta_size` to`add_time_ids` 
                   
                     add_time_ids = self._get_add_time_ids(...)

gnobitab replied:
I'm kind of confused what we should do here...yiyi could you make an example? thanks

gnobitab commented:
> > Our BertModel cannot work in FP16. So to make our pipeline to FP16, we have to make only the transformer to FP16 instead of just pipe = HunyuanDiTPipeline.from_pretrained("XCLiu/HunyuanDiT-0523", transformer=model, torch_dtype=torch.float16)
> 
> Why wouldn't it work, though? Could you provide more details here? In order for the pipeline to operate in `torch.float16` all the components need to be in `torch.float16`, or we have deal with it differently like so:
> 
> https://github.com/huggingface/diffusers/blob/b3d10d6d65a80593627c6738fbeded2f69b5129f/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L1264
> 
> Additionally, could you see if [this](https://github.com/huggingface/diffusers/pull/8240/files#r1614280384) worked so that we can avoid `self.text_encoder.pooler.to_empty(device='cpu')`?

Hi Sayak,
my server was down. I restarted the sever, reinstalled the environment, and the model works with FP16 now... I have no idea why, but the current environment is,
```
torch==2.0.1
transformers==4.41.1
```
I guess the reason is I'm using the latest `transformers` library now. @sayakpaul

sayakpaul commented:
Okay great. It seems like the FP16 problem and also the `to_empty()` problem (solution [here](https://github.com/huggingface/diffusers/pull/8240/files#r1617108976)) are solved now?

gnobitab commented:
I pushed a new version. In this version:
1. Checked and Merged Yiyi's refactor in https://github.com/huggingface/diffusers/commit/b0e0da28c4d0f057824faacb23da6b06dd43a786 (PR https://github.com/huggingface/diffusers/pull/8310)
2. Fixed FP16 and `to_empty` problem.
3. polished the whole codebase following the new comments

The new test file is updated in `test_hunyuan_dit.py`

Thank you for the help! Please review the new version @yiyixuxu @sayakpaul

gnobitab commented:
Note:
(1) Fix FP16: update the `transformers` to `4.41.1` or later.
(2) `to_empty`: https://github.com/huggingface/diffusers/pull/8240/files#r1617108976
```py
from transformers import BertModel
bert_model = BertModel.from_pretrained("XCLiu/HunyuanDiT-0523", add_pooling_layer=True, subfolder="text_encoder")

pipe = HunyuanDiTPipeline.from_pretrained("XCLiu/HunyuanDiT-0523", text_encoder=bert_model, transformer=model, torch_dtype=torch.float32)
pipe.to('cuda')
pipe.save_pretrained("HunyuanDiT-ckpt")

del pipe

pipe = HunyuanDiTPipeline.from_pretrained("HunyuanDiT-ckpt")
pipe.cuda()
```

yiyixuxu commented on src/diffusers/models/embeddings.py:
```diff
@@ -806,6 +920,27 @@ def forward(self, caption):
         return hidden_states
 
 
+# YiYi notes: combine PixArtAlphaTextProjection and HunYuanTextProjection?
```
cc @DN6 here should we use `PixArtAlphaTextProjection`instead? it seems like this projection layer is going to be used a lot

DN6 replied:
Yeah I think we can make it a generic TextProjection class and reuse it.

yiyixuxu commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,499 @@
+        ).to(origin_dtype)
+
+
+class AdaLayerNormShift(nn.Module):
```
@gnobitab, was this a method your team came up with? Is there any reason you chose this over the regular Ada layer norm with both scale and shift? 
I want to understand this because we will need to decide to keep this module here or put it in the embeddings.py so it is easier for other researchers to use it

DN6 replied:
Yeah I think AdaLayerNorm can be reused here as well?

gnobitab replied:
I'll go double check with the team and get back to you later

yiyixuxu replied:
to-do [2]: refactor?

yiyixuxu commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,499 @@
+
+
+@maybe_allow_in_graph
+class HunyuanDiTBlock(nn.Module):
```
I'm putting it here for now but we can make this a generic transformer block or just use the `BasicTransformerBlock` instead

cc @DN6

DN6 replied:
IMO we can leave the block here. BasicTransformerBlock is getting quite large, and there are additional operations here `skip_linear` `skip_norm` , `qk_norm`. The forward pass also introduces new arguments such as `skip` and `image_rotary_emb`

yiyixuxu commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -852,6 +852,7 @@ def load_module(name, value):
         # 7. Load each module in the pipeline
         current_device_map = None
         for name, (library_name, class_name) in logging.tqdm(init_dict.items(), desc="Loading pipeline components..."):
+            print(name, library_name, class_name)
```
```suggestion
```

yiyixuxu commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -869,6 +870,7 @@ def load_module(name, value):
 
             # 7.3 Use passed sub model or load class_name from library_name
             if name in passed_class_obj:
+                print(name)
```
```suggestion
```

yiyixuxu commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,499 @@
+            return (output,)
+        return Transformer2DModelOutput(sample=output)
+
+    def initialize_weights(self):
```
is this needed for training? are we planning to release training script too? 
let's update it if needed, or remove it if not

yiyixuxu reviewed (commented):
thanks!!
I think we can merge this soon

We have tests and doc left - doc can be added in a separate PR if you need more time, but let's quickly add a test, you can use pixart_sigma test as reference https://github.com/huggingface/diffusers/tree/main/tests/pipelines/pixart_sigma

yiyixuxu commented:
cc @DN6 for a final review -  I left a few questions for you, I think they can be addressed in a follow-up PR too

yiyixuxu commented:
@sayakpaul can you help look into, or provide guidance on optimization? i.e. make sure all our optimization methods works on HunyuanDIT!

sayakpaul commented:
> @sayakpaul can you help look into, or provide guidance on optimization? i.e. make sure all our optimization methods works on HunyuanDIT:)

Sure, I will give it a look. Memory optimization-wise, do we want to add feed-forward chunking and QKV fusion? If so, I can add that in a separate commit. Or would you rather handle it? Lwt me know. Cc: @gnobitab too.

yiyixuxu commented:
>memory optimization-wise, do we want to add feed-forward chunking and QKV fusion? If so, I can add that in a separate commit

yes please!

gnobitab commented:
> thanks!! I think we can merge this soon
> 
> We have tests and doc left - doc can be added in a separate PR if you need more time, but let's quickly add a test, you can use pixart_sigma test as reference https://github.com/huggingface/diffusers/tree/main/tests/pipelines/pixart_sigma

@yiyixuxu What kind of docs should I add?

gnobitab commented:
I changed XCLiu/HunayunDiT-0523 and simplified `test_hunyuan_dit.py`

yiyixuxu commented:
@gnobitab 
we can add a page here https://github.com/huggingface/diffusers/tree/main/docs/source/en/api/pipelines
here is pixart sigma doc for your reference
https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/pixart_sigma.md

DN6 commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,499 @@
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+class FP32_Layernorm(nn.LayerNorm):
```
I think it would be better to stick with camelcase here. e.g. FP32LayerNorm

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_8240). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

yiyixuxu commented on src/diffusers/models/normalization.py:
```diff
@@ -176,7 +176,8 @@ def __init__(
 
     def forward(self, x: torch.Tensor, conditioning_embedding: torch.Tensor) -> torch.Tensor:
-        emb = self.linear(self.silu(conditioning_embedding))
+        # covonvert back to the original dtype in case `conditioning_embedding`` is upcasted to float32 (needed for hunyuanDiT)
```
```suggestion
        # convert back to the original dtype in case `conditioning_embedding`` is upcasted to float32 (needed for hunyuanDiT)
```

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,882 @@
+    return noise_cfg
+
+
+class HunyuanDiTPipeline(DiffusionPipeline, TextualInversionLoaderMixin, LoraLoaderMixin, FromSingleFileMixin):
```
```suggestion
class HunyuanDiTPipeline(DiffusionPipeline):
```

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,882 @@
+                Attention mask for the negative prompt. Required when `negative_prompt_embeds` is passed directly.
+            max_sequence_length (`int`, *optional*): maximum sequence length to use for the prompt.
+            text_encoder_index (`int`, *optional*):
+                Index of the text encoder to use. `0` for Bert and `1` for T5.
```
```suggestion
                Index of the text encoder to use. `0` for clip and `1` for T5.
```

yiyixuxu commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,427 @@
+        dropout (`float`, *optional*, defaults to 0.0):
+            The dropout probability to use.
+        cross_attention_dim (`int`, *optional*):
+            The number of dimension in the bert text embedding.
```
```suggestion
            The number of dimension in the clip text embedding.
```

yiyixuxu commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,427 @@
+        pooled_projection_dim (`int`, *optional*):
+            The size of the pooled projection.
+        text_len (`int`, *optional*):
+            The length of the bert text embedding.
```
```suggestion
            The length of the clip text embedding.
```

yiyixuxu commented on src/diffusers/models/embeddings.py:
```diff
@@ -507,6 +626,88 @@ def forward(self, timestep, class_labels, hidden_dtype=None):
         return conditioning
 
 
+class HunyuanDiTAttentionPool(nn.Module):
```
to-do[1] (low prior): see if we can consolidate this class with https://github.com/huggingface/diffusers/blob/bc108e15333cb0e8a092647320cbb4d70d6d0f03/src/diffusers/models/embeddings.py#L588 and https://github.com/huggingface/diffusers/blob/bc108e15333cb0e8a092647320cbb4d70d6d0f03/src/diffusers/models/unets/unet_kandinsky3.py#L483
can ask community for help

yiyixuxu commented on src/diffusers/models/transformers/hunyuan_transformer_2d.py:
```diff
@@ -0,0 +1,427 @@
+        return hidden_states
+
+
+class HunyuanDiT2DModel(ModelMixin, ConfigMixin):
```
to-do[3] optimization! cc @sayakpaul

yiyixuxu commented on src/diffusers/pipelines/hunyuandit/pipeline_hunyuandit.py:
```diff
@@ -0,0 +1,881 @@
+        text_encoder: BertModel,
+        tokenizer: BertTokenizer,
+        transformer: HunyuanDiT2DModel,
+        scheduler: DDPMScheduler,
```
[to-do 3] Does hunyuan DiT work with other schedulers? @gnobitab 
if so, we need to change the type here, and also let's update the doc example with a more efficient scheduler 
```suggestion
        scheduler: KarrasDiffusionSchedulers,
```

yiyixuxu commented on tests/pipelines/hunyuan_dit/test_hunyuan_dit.py:
```diff
@@ -0,0 +1,266 @@
+enable_full_determinism()
+
+
+class HunyuanDiTPipelineFastTests(PipelineTesterMixin, unittest.TestCase):
```
to-do[4] 

I added s simple fast test and slow test. @gnobitab can we add more tests?  see examples https://github.com/huggingface/diffusers/blob/bc108e15333cb0e8a092647320cbb4d70d6d0f03/tests/pipelines/pixart_sigma/test_pixart.py 

we help with this too but we prefer authors to add tests and make sure the current implementation is correct!

yiyixuxu commented on tests/pipelines/hunyuan_dit/test_hunyuan_dit.py:
```diff
@@ -0,0 +1,266 @@
+        max_diff = np.abs(image_slice.flatten() - expected_slice).max()
+        self.assertLessEqual(max_diff, 1e-3)
+
+    def test_sequential_cpu_offload_forward_pass(self):
```
to-do[5] (slow prior) the sequential cpu offload does not work, can look into this. but not a priority since the other offloading method works and is way more popular cc @sayakpaul

yiyixuxu commented:
merging now! 

I left a bunch to-dos in the comments too 

for @sayakpaul 
can you help with optimization? make sure `torch.compile` works and all the other meaningful optimization method works here 

for @gnobitab 
1. add doc 
2. add more tests if possible
3. I left a few questions for you about the ada layer norm and the scheduler 
4. move the checkpoint to the official account. 
5. do we need a conversion script? cc @sayakpaul too. is the diffuser checkpoint the main checkpoint that is going to be used by the community? 

here is the script that works now for the checkpoint in my PR to your repo https://huggingface.co/XCLiu/HunyuanDiT-0523/discussions/2
```python

# integration test (hunyuan dit)
import torch

from diffusers import HunyuanDiTPipeline


device = "cuda"
dtype = torch.float16

repo = "XCLiu/HunyuanDiT-0523"

pipe = HunyuanDiTPipeline.from_pretrained(repo, revision="refs/pr/2", torch_dtype=dtype)
pipe.enable_model_cpu_offload()

### NOTE: HunyuanDiT supports both Chinese and English inputs
prompt = "一个宇航员在骑马"
#prompt = "An astronaut riding a horse"
generator=torch.Generator(device="cuda").manual_seed(0)
image = pipe(height=1024, width=1024, prompt=prompt, generator=generator).images[0]

image.save("yiyi_test_out.png")
```

and this is the script I used to convert the current checkpoint in `"XCLiu/HunyuanDiT-0523"`
basically I did these two changes:
1. changed the module/folder name for t5 text_encoder and tokenizer to `text_encoder_2` and `tokenizer_2`
2. updated the state dict for transformers because I refactored the embedding layers for the extra conditions 

```python
import torch
from huggingface_hub import hf_hub_download

from diffusers import HunyuanDiTPipeline, HunyuanDiT2DModel
from transformers import T5EncoderModel, T5Tokenizer

import safetensors.torch

device = "cuda"
dtype = torch.float32

repo = "XCLiu/HunyuanDiT-0523"

tokenizer_2 = T5Tokenizer.from_pretrained(repo, subfolder = "tokenizer_t5")
text_encoder_2 = T5EncoderModel.from_pretrained(repo, subfolder = "embedder_t5", torch_dtype=dtype)

model_config = HunyuanDiT2DModel.load_config("XCLiu/HunyuanDiT-0523", subfolder="transformer")
model = HunyuanDiT2DModel.from_config(model_config).to(device)

ckpt_path = hf_hub_download(
    "XCLiu/HunyuanDiT-0523",
    filename ="diffusion_pytorch_model.safetensors",
    subfolder="transformer",
)

state_dict = safetensors.torch.load_file(ckpt_path)

prefix = "time_extra_emb."

# time_embedding.linear_1 -> timestep_embedder.linear_1 
state_dict[f"{prefix}timestep_embedder.linear_1.weight"] = state_dict["time_embedding.linear_1.weight"]
state_dict[f"{prefix}timestep_embedder.linear_1.bias"] = state_dict["time_embedding.linear_1.bias"]
state_dict.pop("time_embedding.linear_1.weight")
state_dict.pop("time_embedding.linear_1.bias")

# time_embedding.linear_2 -> timestep_embedder.linear_2
state_dict[f"{prefix}timestep_embedder.linear_2.weight"] = state_dict["time_embedding.linear_2.weight"]
state_dict[f"{prefix}timestep_embedder.linear_2.bias"] = state_dict["time_embedding.linear_2.bias"]
state_dict.pop("time_embedding.linear_2.weight")
state_dict.pop("time_embedding.linear_2.bias")

# pooler.positional_embedding
state_dict[f"{prefix}pooler.positional_embedding"] = state_dict["pooler.positional_embedding"]
state_dict.pop("pooler.positional_embedding")

# pooler.k_proj
state_dict[f"{prefix}pooler.k_proj.weight"] = state_dict["pooler.k_proj.weight"]
state_dict[f"{prefix}pooler.k_proj.bias"] = state_dict["pooler.k_proj.bias"]
state_dict.pop("pooler.k_proj.weight")
state_dict.pop("pooler.k_proj.bias")

#pooler.q_proj
state_dict[f"{prefix}pooler.q_proj.weight"] = state_dict["pooler.q_proj.weight"]
state_dict[f"{prefix}pooler.q_proj.bias"] = state_dict["pooler.q_proj.bias"]
state_dict.pop("pooler.q_proj.weight")
state_dict.pop("pooler.q_proj.bias")

#  pooler.v_proj
state_dict[f"{prefix}pooler.v_proj.weight"] = state_dict["pooler.v_proj.weight"]
state_dict[f"{prefix}pooler.v_proj.bias"] = state_dict["pooler.v_proj.bias"]
state_dict.pop("pooler.v_proj.weight")
state_dict.pop("pooler.v_proj.bias")

# pooler.c_proj
state_dict[f"{prefix}pooler.c_proj.weight"] = state_dict["pooler.c_proj.weight"]
state_dict[f"{prefix}pooler.c_proj.bias"] = state_dict["pooler.c_proj.bias"]
state_dict.pop("pooler.c_proj.weight")
state_dict.pop("pooler.c_proj.bias")

# style_embedder.weight
state_dict[f"{prefix}style_embedder.weight"] = state_dict["style_embedder.weight"]
state_dict.pop("style_embedder.weight")

# extra_embedder.linear_1
state_dict[f"{prefix}extra_embedder.linear_1.weight"] = state_dict["extra_embedder.linear_1.weight"]
state_dict[f"{prefix}extra_embedder.linear_1.bias"] = state_dict["extra_embedder.linear_1.bias"]
state_dict.pop("extra_embedder.linear_1.weight")
state_dict.pop("extra_embedder.linear_1.bias")

# extra_embedder.linear_2
state_dict[f"{prefix}extra_embedder.linear_2.weight"] = state_dict["extra_embedder.linear_2.weight"]
state_dict[f"{prefix}extra_embedder.linear_2.bias"] = state_dict["extra_embedder.linear_2.bias"]
state_dict.pop("extra_embedder.linear_2.weight")
state_dict.pop("extra_embedder.linear_2.bias")

model.load_state_dict(state_dict)
model.to(dtype)

pipe = HunyuanDiTPipeline.from_pretrained(
    repo, 
    tokenizer_2 = tokenizer_2,
    text_encoder_2 = text_encoder_2,
    transformer = model,
    torch_dtype=dtype)

```

gnobitab commented:
@yiyixuxu 

Thanks for merging!

Reply to your TODOs:
1. Doc added here: https://moon-ci-docs.huggingface.co/docs/diffusers/pr_8383/en/api/pipelines/hunyuandit

2. I manually tested and verified the results. The current version has the same output as the original model. I will try to add more rigorous tests later.

3. (1) **AdaLayerNormShift**: I asked the team member and they were following the practice in SDXL: See 
https://github.com/huggingface/diffusers/blob/413604405fddb4692a8e9a9a9fb6c353d22881ea/src/diffusers/models/resnet.py#L343 (L343 - L351). I notice the place of `self.norm` in HunyuanDiT is different from SDXL, but I think it is just a small mistake. As far as I learned, it is not a scientific innovation and I suggest to keep it inside `hunyuan_transformer_2d.py`

(2) **Scheduler**: I tested several fast samplers. From my test, I think it is safe to switch from ``DDPMScheduler` to `DDIMScheduler`. Other schedulers seem to be fine but I would like to leave that for the community.

4 and 5: I merged your PR in `XCLiu/HunyuanDiT-0523` and saved a new checkpoint for the diffusers pipeline. We are moving the new checkpoint to the official account after going through some internal checks. The name will be `Tencent-Hunyuan/HunyuanDiT-Diffusers`. Let's change the example doc in the pipeline file when it's officially online.
